src/uct/cuda/base/cuda_md.c

```diff
@@ -17,42 +17,133 @@
 #include <cuda.h>
 
 
+static ucs_status_t uct_cuda_base_get_sys_dev(CUdevice cuda_device,
+                                              ucs_sys_device_t *sys_dev_p)
+{
+    ucs_sys_bus_id_t bus_id;
+    CUresult cu_err;
+    int attrib;
+
+    /* PCI domain id */
+    cu_err = cuDeviceGetAttribute(&attrib, CU_DEVICE_ATTRIBUTE_PCI_DOMAIN_ID,
+                                  cuda_device);
+    if (cu_err != CUDA_SUCCESS) {
+         return UCS_ERR_IO_ERROR;
+    }
+    bus_id.domain = (uint16_t)attrib;
+
+    /* PCI bus id */
+    cu_err = cuDeviceGetAttribute(&attrib, CU_DEVICE_ATTRIBUTE_PCI_BUS_ID,
+                                  cuda_device);
+    if (cu_err != CUDA_SUCCESS) {
+         return UCS_ERR_IO_ERROR;
+    }
+    bus_id.bus = (uint8_t)attrib;
+
+    /* PCI slot id */
+    cu_err = cuDeviceGetAttribute(&attrib, CU_DEVICE_ATTRIBUTE_PCI_DEVICE_ID,
+                                  cuda_device);
+    if (cu_err != CUDA_SUCCESS) {
+         return UCS_ERR_IO_ERROR;
+    }
+    bus_id.slot = (uint8_t)attrib;
+
+    /* Function - always 0 */
+    bus_id.function = 0;
+
+    return ucs_topo_find_device_by_bus_id(&bus_id, sys_dev_p);
+}
+
+
 UCS_PROFILE_FUNC(ucs_status_t, uct_cuda_base_detect_memory_type,
                  (md, addr, length, mem_type_p),
                  uct_md_h md, const void *addr, size_t length,
                  ucs_memory_type_t *mem_type_p)
 {
-    CUmemorytype memType = (CUmemorytype)0;
-    uint32_t isManaged   = 0;
-    unsigned value       = 1;
-    void *attrdata[] = {(void *)&memType, (void *)&isManaged};
-    CUpointer_attribute attributes[2] = {CU_POINTER_ATTRIBUTE_MEMORY_TYPE,
-                                         CU_POINTER_ATTRIBUTE_IS_MANAGED};
-    CUresult cu_err;
+    uct_md_mem_attr_t mem_attr;
+    ucs_status_t status;
+
+    mem_attr.field_mask = UCT_MD_MEM_ATTR_FIELD_MEM_TYPE;
+    status = uct_cuda_base_mem_query(md, addr, length, &mem_attr);
+    if (status != UCS_OK) {
+        return status;
+    }
+
+    *mem_type_p = mem_attr.mem_type;
+    return UCS_OK;
+}
+
+UCS_PROFILE_FUNC(ucs_status_t, uct_cuda_base_mem_query,
+                 (md, address, length, mem_attr),
+                 uct_md_h md, const void *address, size_t length,
+                 uct_md_mem_attr_t *mem_attr)
+{
+    CUmemorytype cuda_mem_mype = (CUmemorytype)0;
+    uint32_t is_managed        = 0;
+    unsigned value             = 1;
+    CUdevice cuda_device       = -1;
+    CUpointer_attribute attr_type[3];
```

Comment:
define a local constant for `3` that is used twice

Reply:
fixed